google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/OpenTelemetryUtil.java

```diff
@@ -0,0 +1,198 @@
+          message.addPublishStartEvent();
+        }
+      }
+      return publishRpcSpan;
+    }
+    return null;
+  }
+
+  /** Ends the given publish RPC span if it exists. */
+  public static final void endPublishRpcSpan(
+      Span publishRpcSpan, boolean enableOpenTelemetryTracing) {
+    if (enableOpenTelemetryTracing && publishRpcSpan != null) {
+      publishRpcSpan.end();
+    }
+  }
+
+  /**
+   * Sets an error status and records an exception when an exception is thrown when publishing the
+   * message batch.
+   */
+  public static final void setPublishRpcSpanException(
+      Span publishRpcSpan, Throwable t, boolean enableOpenTelemetryTracing) {
+    if (enableOpenTelemetryTracing && publishRpcSpan != null) {
+      publishRpcSpan.setStatus(StatusCode.ERROR, "Exception thrown on publish RPC.");
+      publishRpcSpan.recordException(t);
+      publishRpcSpan.end();
+    }
+  }
+
+  /**
+   * Creates, starts, and returns spans for ModAck, Nack, and Ack RPC requests. Bi-directional links
+   * to parent subscribe span for sampled messages are added.
+   */
+  public static final Span startSubscribeRpcSpan(
+      Tracer tracer,
+      String subscription,
+      String rpcOperation,
+      List<PubsubMessageWrapper> messages,
+      int ackDeadline,
+      boolean isReceiptModack,
+      boolean enableOpenTelemetryTracing) {
+    if (enableOpenTelemetryTracing && tracer != null) {
+      String codeFunction =
+          rpcOperation == "ack"
+              ? "StreamingSubscriberConnection.sendAckOperations"
+              : "StreamingSubscriberConnection.sendModAckOperations";
+      SubscriptionName subscriptionName = SubscriptionName.parse(subscription);
+      AttributesBuilder attributesBuilder =
+          createCommonSpanAttributesBuilder(
+                  subscriptionName.getSubscription(),
+                  subscriptionName.getProject(),
+                  codeFunction,
+                  rpcOperation)
+              .put(SemanticAttributes.MESSAGING_BATCH_MESSAGE_COUNT, messages.size());
+
+      // Ack deadline and receipt modack are specific to the modack operation
+      if (rpcOperation == "modack") {
+        attributesBuilder
+            .put(ACK_DEADLINE_ATTR_KEY, ackDeadline)
+            .put(RECEIPT_MODACK_ATTR_KEY, isReceiptModack);
+      }
+
+      Span rpcSpan =
+          tracer
+              .spanBuilder(subscriptionName.getSubscription() + " " + rpcOperation)
+              .setSpanKind(SpanKind.CLIENT)
+              .setAllAttributes(attributesBuilder.build())
+              .startSpan();
+
+      for (PubsubMessageWrapper message : messages) {
+        if (message.getSubscriberSpan().getSpanContext().isSampled()) {
+          Attributes linkAttributes =
+              Attributes.builder()
+                  .put(SemanticAttributes.MESSAGING_OPERATION, rpcOperation)
+                  .build();
+          rpcSpan.addLink(message.getSubscriberSpan().getSpanContext(), linkAttributes);
```

Comment:
Same comment as above: call `addLink` as part of span creation when possible.

Reply:
Did the same as for the `publishRpcSpan`.